### DISCUSSION TITLE
How to deploy Nebari on existing Azure Kubernetes cluster

### DISCUSSION THREAD

**Discussion by Traychy** [nebari-dev/nebari#2118] [Q&A]
Hey all, I am working on a project which needs to test Nebari on AKS which already exists. I saw that there is a specific article in the Nebari's official documentation for deploying on pre-existing kubernetes cluster  - https://www.nebari.dev/docs/how-tos/nebari-kubernetes/ 
In the article only AWS is mentioned so I was wondering if there is any difference for Azure and if yes what is it? Should we wait for updated version of the article including the Azure/GCP deployment on pre-existing kubernetes clusters?

**Comment by rsignell** [comment]
@Traychy thanks for raising the issue here!    I don't know much about Azure either, but @Adam-D-Lewis provided this info to me offline: 
>  Nebari runs on a managed kubernetes cluster (AKS on azure).  Azure requires that a new VNet is created during the creation of the AKS cluster so that won't work if you require pre existing Vnets.  Additionally Nebari will try to create several Vnets by default.  I believe you can use pre-existing vnets for all the vnets ( except for the AKS one) using Terraform overrides.  Additionally, if you are deploying a private AKS cluster (aks nodes don't have a public IP address), then you'll need to deploy from a VM on Azure that can still reach the nodes.

**Reply by Traychy** [reply]
Thank you for the information. The AKS is connected to a subnet which is part of private vnet - would this be a problem ? @Adam-D-Lewis  are there any steps that should be followed for deploying Nebari on pre-existing AKS? 
Note: VM that has access to the nodes is an option on our end.

**Reply by Adam-D-Lewis** [reply]
> The AKS is connected to a subnet which is part of private vnet - would this be a problem?  

I think the install would be similar to an install on a private vnet.

> @Adam-D-Lewis are there any steps that should be followed for deploying Nebari on pre-existing AKS?

Although we only have a guide for AWS, it may be helpful for you - https://www.nebari.dev/docs/how-tos/nebari-kubernetes.  If you go this route and are able to get it working on Azure, any contributions to the docs would be appreciated :)

**Reply by Traychy** [reply]
Thank you for the update @Adam-D-Lewis. Installation on a private vnet - can you refer to an article as part of the documentation?

**Reply by rsignell** [reply]
@Traychy , I'm not sure what you mean here, but I think it would be valuable to just follow the AWS instructions and translate the concepts from AWS to Azure.   Let's just take it one step at a time: Looking at https://www.nebari.dev/docs/how-tos/nebari-kubernetes/#evaluating-the-infrastructure, do you have an existing EKS cluster that runs an app?   Does it have node groups?  If not, you will need to create them.  The instructions say: 
> Before proceeding, ensure that the subnets can "[automatically assign public IP addresses to instances launched into it](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-ip-addressing.html#subnet-public-ip)" and that there exists an [Identity and Access Management (IAM)](https://docs.aws.amazon.com/eks/latest/userguide/create-node-role.html) role with the following permissions:
AmazonEKSWorkerNodePolicy
AmazonEC2ContainerRegistryReadOnly
AmazonEKS_CNI_Policy

I would think this should translate into a Azure service principal with appropriate permissions, etc.

**Reply by Traychy** [reply]
Hey @rsignell , I have asked for article because I was confused what @Adam-D-Lewis  has mentioned ' install on a private vnet ' but also as prerequisites public IP address is required. I thought that public IPs are not needed because I saw an article for local deployment - https://www.nebari.dev/docs/how-tos/nebari-local/ so maybe combination between two could help.

**Reply by Traychy** [reply]
@Adam-D-Lewis , May I ask if public IPs are needed for Nebari to operate or only during the deployment for downloading dependencies and etc?

**Reply by rsignell** [reply]
@Adam-D-Lewis I'm pretty sure it's the latter, right?    If so, I'm sure we could work around that.  

Also, I believe you said you helped one of your clients deal with similar issues on Azure and might have a sample config that could be useful to us?

**Reply by Adam-D-Lewis** [reply]
Yeah, As long as the machine deploying nebari can reach the nodes, then the nodes don't need a public IP address.  I can share an example configuration that may help, but I'm traveling at the moment and won't have access to it until next week.

**Reply by rsignell** [reply]
Okay, thanks @Adam-D-Lewis !  Have a great weekend!

**Reply by Traychy** [reply]
We get to a point where Nebari is reaching the cluster and passing stage2 but it is failing on stage3-kubernetes-initialize with this error 'TypeError: Object of type KeyValueDict is not JSON serializable'. Tried to look at which exact module is failing but still no success. Namespace and traefik CRDs are deployed so I am wondering if the script is complaining about the node-selectors or some of the modules in stage3. Can you guide us in any direction ?

**Reply by Traychy** [reply]
It appeared that this is a bug in the code. We managed to fix it on our end and there is a pull request for it on your repo. Other than that, there is an issue with this - Error: admission webhook "validation.gatekeeper.sh" denied the request: [azurepolicy-k8sazurev1loadbalancernopublic-601d80ac4cae0639f9c6] Load Balancers should not have public IPs. azure-load-balancer-internal annotation is required for nebari-traefik-ingress. Ingress section is configured  in the nebari-config.yaml with the needed information but still the same error.

**Comment by dharhas** [comment]
@Traychy I'm converting this from a discussion to an issue so we can track it better.

**Reply by rsignell** [reply]
https://github.com/nebari-dev/nebari/issues/2131
